proposals/0133-syscall-get-epoch-stake.md

```diff
@@ -0,0 +1,96 @@
+---
+simd: '0133'
+title: Syscall Get-Epoch-Stake
+authors:
+  - Joe Caulfield (Anza)
+category: Standard
+type: Core
+status: Draft
+created: 2024-03-25
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+A syscall to retrieve a vote account's delegated stake for the current epoch.
+
+## Motivation
+
+Currently, on-chain programs have no knowledge of the current epoch's stake and
+how much active stake is delegated to any vote account.
+
+If this data were available for querying by on-chain programs, it would unblock
+many use cases, such as validator governance and secondary consensus mechanisms,
+that were previously not possible on Solana.
+
+Additionally, this would enable the Feature Gate program defined in
+[SIMD 0089](./0089-programify-feature-gate.md) to tally vote account stake in
+support for a pending feature gate.
+
+## Alternatives Considered
+
+[SIMD 0056](https://github.com/solana-foundation/solana-improvement-documents/pull/56)
+proposes using an on-chain sysvar account to store all of the current epoch's
+stake.
+
+Because account data is finite, using a sysvar account to store even just the 
+current epoch's stake limits the number of entries that can be stored. The
+amount of validators in one Solana cluster could surpass this number in the
+future.
+
+Exposing epoch-stake information through a syscall avoids this account maximum
+size constraint. While the syscall approach does not offer the easy off-chain
+retrieval of a sysvar, there are existing ways to get epoch-stake data off
+chain. The priority of a new design should be making the data available to
+on-chain programs.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
+
+The specification for the proposed syscall is as follows:
+
+```c
+/**
+ * Retrieves the total active stake delegated to a vote account for the current
+ * epoch.
+ *
+ * @param var_addr      VM memory address to copy the retrieved data to.
+ * @param vote_address  A pointer to 32 bytes representing the vote address.
+ * @return              A 64-bit unsigned integer error code:
+ *                        - 0 if the operation is successful.
+ *                        - Non-zero error code.
+ *
+ * If the operation is not successful, data will not be written to the
+ * provided VM memory address.
+ */
+uint64_t sol_get_epoch_stake(
+  /* r1 */ uint8_t *    var_addr,
+  /* r2 */ void const * vote_address,
+);
```

Comment:
It does not break any existing patterns for existing syscalls, since they all seem to do
things differently.

For example, many cryptographic syscalls - such as `sol_curve_multiscalar_mul` -
use `1` to indicate an `Option::None` response from curve calculation.

https://github.com/anza-xyz/agave/blob/b1e17998b64203fe62f52efe3fe5ba87706ed446/programs/bpf_loader/src/syscalls/mod.rs#L1180-L1189

https://github.com/anza-xyz/agave/blob/b1e17998b64203fe62f52efe3fe5ba87706ed446/zk-token-sdk/src/curve25519/ristretto.rs#L227-L242

While others - such as `sol_secp256k1_recover` - use the integer value to map specific errors,
thrown intentionally based on computation. Note these are separate from errors such as memory
access or unavailable data (like in sysvars).

https://github.com/anza-xyz/agave/blob/b1e17998b64203fe62f52efe3fe5ba87706ed446/programs/bpf_loader/src/syscalls/mod.rs#L842-L853

https://github.com/anza-xyz/agave/blob/b1e17998b64203fe62f52efe3fe5ba87706ed446/sdk/program/src/secp256k1_recover.rs#L409-L422

Finally, a handful of syscalls - such as `sol_remaining_compute_units` simply return the
value in the resulting `u64`, as you've mentioned.

https://github.com/anza-xyz/agave/blob/b1e17998b64203fe62f52efe3fe5ba87706ed446/programs/bpf_loader/src/syscalls/mod.rs#L1735-L1753

Reply:
I've added a commit changing this syscall to return the data directly, in the interest of saving a memory access.